elasticsearch_acls.go

```diff
@@ -0,0 +1,258 @@
+		for _, permission := range elasticsearchPermissins {
+			if p == permission {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	for _, acl := range acls {
+		// check ACL permission
+		for _, rule := range acl.Rules {
+			if !isValidPermission(rule.Permission) {
+				return errors.New("invalid ACL permission :" + rule.Permission)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Create creates new Elasticsearch ACLs config
+func (h *ElasticSearchACLsHandler) Create(project, service string, req ElasticsearchACLRequest) (*ElasticSearchACLResponse, error) {
+	// check is ACL is empty
+	for _, acl := range req.ElasticSearchACLConfig.ACLs {
+		if acl.Username == "" || len(acl.Rules) == 0 {
+			return nil, errors.New("cannot create an ACL with empty username or list of rules")
+		}
+	}
+
+	req.ElasticSearchACLConfig.Enabled = true
+
+	return h.createOrUpdate(project, service, req, true)
+}
+
+// Update updates Elasticsearch ACL config
+func (h *ElasticSearchACLsHandler) Update(project, service string, req ElasticsearchACLRequest) (*ElasticSearchACLResponse, error) {
+	return h.createOrUpdate(project, service, req, false)
+}
+
+// createOrUpdate sends POST(create) or PUT(update) requests to Aiven API
+func (h *ElasticSearchACLsHandler) createOrUpdate(project, service string, req ElasticsearchACLRequest, create bool) (*ElasticSearchACLResponse, error) {
+	var bts []byte
+	var err error
+
+	// validate ACLs
+	err = h.checkACL(req.ElasticSearchACLConfig.ACLs)
+	if err != nil {
+		return nil, err
+	}
+
+	// POST or PUT request
+	path := buildPath("project", project, "service", service, "elasticsearch", "acl")
+	if create {
+		bts, err = h.client.doPostRequest(path, req)
+	} else {
+		bts, err = h.client.doPutRequest(path, req)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var rsp *ElasticSearchACLResponse
+	if err := json.Unmarshal(bts, &rsp); err != nil {
+		return nil, err
+	}
+
+	return h.checkResponse(rsp)
+}
+
+// checkResponse checks if response fom Aiven API contains any errors
+func (h *ElasticSearchACLsHandler) checkResponse(response *ElasticSearchACLResponse) (*ElasticSearchACLResponse, error) {
+	// response cannot be empty
+	if response == nil {
+		return nil, ErrNoResponseData
+	}
+
+	// check API response errors
+	if response.Errors != nil && len(response.Errors) != 0 {
+		return nil, errors.New(response.Message)
+	}
+
+	return response, nil
+}
+
+// List lists all existing Elasticsearch ACLs config
+func (h *ElasticSearchACLsHandler) List(project, service string) (*ElasticSearchACLResponse, error) {
+	path := buildPath("project", project, "service", service, "elasticsearch", "acl")
+	bts, err := h.client.doGetRequest(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var rsp *ElasticSearchACLResponse
+	if err := json.Unmarshal(bts, &rsp); err != nil {
+		return nil, err
+	}
+
+	return h.checkResponse(rsp)
+}
+
+// Add function adds one new ACL record to already existing Elasticsearch ACL configuration,
```

Comment:
I'm not quite sure how these should be done. As there is now AclConfig containing a list of Acl's identified by user each having a list of AclRules, maybe structuring around those could make it clearer? E.g in ElasticsearchAclConfig have the methods to add, remove, update ElasticsearchAcls and in ElasticsearchAcls have methods to add/delete rules?

Reply:
I see your point, I will experiment with these things

Reply:
Did some refactoring, please have a look